Assingment_Tving/Assingment_Tving/Global/Extensions/UIViewController+Extension.swift

```diff
@@ -0,0 +1,45 @@
+//
+//  UIViewController+Extension.swift
+//  Assingment_Tving
+//
+//  Created by 이성민 on 2023/04/12.
+//
+
+import UIKit
+
+extension UIViewController {
+    
+    // MARK: - navigation
+    
+    func makeNavigationBarButton<T: UIView>(with view: T) -> UIBarButtonItem {
+        return UIBarButtonItem(customView: view)
+    }
+    
+    func makeNavigationBarItemArea(with button: UIButton) -> UIView {
+        let offsetView = UIView(frame: CGRect(x: 0, y: 0, width: 44, height: 44))
+        offsetView.addSubview(button)
+        return offsetView
+    }
+    
+    @objc
+    func popViewController() {
+        self.navigationController?.popViewController(animated: true)
+    }
+    
+    // MARK: - keyboard gesture
+    
+    func setKeyboard() {
+        let tap = UITapGestureRecognizer(target: self, action: #selector(endEditing))
+        tap.delegate = self
+        view.addGestureRecognizer(tap)
+    }
```

Comment:
`tap.delegate = self` 안하면 어떻게 되나여?
왜 하셨는지 혹은 안하면 어떤 동작이 안되는지 궁금합니다

Reply:
오우 예리하시군요 후후
솔직히 말하자면 이전 플젝에서 긁어온거긴 한데 ㅎㅎ,,,
저도 궁금해져서 찾아보니까 일단 현재에는 문제가 없습니다..!

---

제가 긁어온 플젝에서는 UIViewController+Extension에 아래와 같은 코드들도 추가가 되어 있는데요,
하나씩 테스트 해보면서 어떤 용도인지 찾아봤숩니다

<img width="952" alt="image" src="https://user-images.githubusercontent.com/72431640/232321779-785a7c30-4eff-40a0-a71c-1d6fedaa0438.png">

---

먼저 위에 코드는 구글링 해 보니까 NavigationBar 를 숨겼을 때 발생하는(SwiftUI 기준 / UIKit 에서는 다른 상황일수도?!)
스와이프 시 이전 View로 돌아가기가 안되는 상황을 해결하기 위한 코드라고 합니다!

NavigationController에 속한 ViewController의 개수가 2개 이상일 때 스와이프 하면 터치를 인식해서,
뒤로가기도 가능케 하는거죠!

---

아래 코드는 설명하기 전에 먼저 상황 가정을 하겠습니다.. (32149184192번째 코멘트 수정중...)

<br>

**[상황]**
석우와 성민이는 `Button`과 `TextField`가 모두 있는 `Modal`을 만들고 있다.
`TextField`를 터치 시 키보드가 올라오고, 그 외의 영역을 터치하면 키보드가 내려가게 하려고 한다
하지만 `Button`을 누르면 `Modal`이 내려가는 상황이다

`Button` 의 터치가 우선시될까 키보드 내리는 터치가 우선시 될까?

<br>

**[시각자료]**

https://user-images.githubusercontent.com/72431640/232327376-7f4eecee-831d-4fed-9ae1-8bc5af362049.mp4


https://user-images.githubusercontent.com/72431640/232327379-f8a9e9f3-3b16-4782-b613-b399f56763f2.mp4



<br>
<br>

**[문제의 코드들]**
1번째 문제의 코드
```swift
let tap = UITapGestureRecognizer(target: self, action: #selector(endEditingView))
tap.cancelsTouchesInView = false // <---
```

<br>

2번째 문제의 코드
```swift
extension UIViewController: UIGestureRecognizerDelegate {
   public func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldReceive touch: UITouch) -> Bool {
      guard touch.view is UIButton else { return true }
      return false
   }
}
```

<br>

**[여러가지 테스트 결과]**
1, 2번째 문제의 코드 모두 주석처리 -> endEditingView 인식 안되고, Button 터치가 인식되어 Modal dismiss
1 번째 코드 사용, 2번째 코드 주석 -> endEditingView 인식되어 키보드 내려감, Button 터치 인식되지 않음
1, 2 번째 코드 모두 사용 -> endEditingView 인식 안되고, Button 터치가 인식되어 Modal dismiss

<br>

**[분석]**
cancelsTouchesInView 를 찾아보니까
> // default is YES. causes touchesCancelled:withEvent: or pressesCancelled:withEvent: to be sent to the view for all touches or presses recognized as part of this gesture immediately before the action method is called.

해석하자면, 디폴트 값은 true 이다, 이 제스쳐의 일부로 간주되는 터치나 프레스를 갖고 있는 모든 view에 action 메소드가 호출되기 전에 touchesCancelled 나 pressesCancelled 메소드가 먼저 호출되도록 한다 랍니다.......(?)

얘가 true 였다가 false 로 바꾸니까 `endEditingView` 만 실행되는걸로 보아 
- true 이면 VC에 포함되어 있는 View에 다른 tapGesture가 있을 때 현재 '나의' tapGesture는 cancel 한다
- false 면 VC에 포함되어 있는 View에 다른 tapGesture가 있더라도 현재 '나의' tapGesture는 cancel 하지 않는다
정도로 생각이 듭니다... 맞는지는 모름... @parkikbum  도와줘요 애니메이션 마스터...

.

공부하다 보니 두 번째 코드가 뭔지는 파악이 됐숩니다

두 번째 문제의 함수를 `UIGestureDelegate` 를 뜯어보니까
> // called before touchesBegan:withEvent: is called on the gesture recognizer for a new touch. return NO to prevent the gesture recognizer from seeing this touch

gesture recognizer 가 이 터치를 인지하지 못하게 하려면 false 를 return 해라

작성한 코드에서도 보면, `guard touch.view is UIButton else { return true } return false`인데,
touch 한 view 가 UIButton 이면 false 
-> touch 한 view 가 UIButton 이면 이 터치를 인지 못하게 해라
-> Button 터치 시 Button의 터치만 인식되게 한다, 즉, modal dismiss 만 발생한다
로 분석이 되었습니다..

---

아무튼 그래서 왜 이렇게까지 뜯어봤는지는... 모르겠는데... 
결론을 보자면 `tap.delegate = self` 는 아직 없어도 되는 코드였고,
코드를 쓰기 전에 그 필요성을 느끼고 충분히 아는 상태에서 사용하자....
그리고 긁어올 때에도 마찬가지.. 확실하게 알고 긁어오자.......

이상 시험공부 하기 싫어서 아요 공부만 주구장창 한 이성민이었습니다...

Reply:
와우우... ㅋㅎㅋㅎ 셤공부가 하기 싫을 때 아요공부가 재밌긴합니다 ㅋㅎ
제가 알기론 한 터치영역에 respoder이 두개 이상일때에는 `역순 깊이 우선 순회 알고리즘` 을 채택한다고 합니다..
터치 이벤트 관련 좋은 글 첨부해드립미당 (진짜 흥미로움! 꼭 셤 끝나고 보세요 ㅋ)

[Hit Testing](https://baked-corn.tistory.com/128)
[Responder Chain, Touch Event](https://baked-corn.tistory.com/129)